src/django_upgrade/fixers/compatibility_imports.py

```diff
@@ -0,0 +1,128 @@
+  Undocumented change
+"""
+from __future__ import annotations
+
+import ast
+from collections import defaultdict
+from functools import lru_cache
+from functools import partial
+from typing import Iterable
+from typing import Mapping
+
+from tokenize_rt import Offset
+
+from django_upgrade.ast import ast_start_offset
+from django_upgrade.ast import is_rewritable_import_from
+from django_upgrade.data import Fixer
+from django_upgrade.data import State
+from django_upgrade.data import TokenFunc
+from django_upgrade.tokens import update_import_modules
+
+fixer = Fixer(
+    __name__,
+    min_version=(0, 0),
+)
+
+REPLACEMENTS_EXACT = {
+    (1, 9): {
+        "django.forms.forms": {
+            "pretty_name": "django.forms.utils",
+            "BoundField": "django.forms.boundfield",
+        },
+    },
+    (1, 11): {
+        "django.db.models.fields": {
+            "FieldDoesNotExist": "django.core.exceptions",
+        },
+        "django.db.models.query": {
+            "EmptyResultSet": "django.core.exceptions",
+        },
+        "django.db.models.sql": {
+            "EmptyResultSet": "django.core.exceptions",
+        },
+        "django.db.models.sql.datastructures": {
+            "EmptyResultSet": "django.core.exceptions",
+        },
+    },
+    (2, 0): {"django.utils.functional": {"lru_cache": "functools"}},
+    (3, 1): {
+        "django.contrib.postgres.forms": {
+            "JSONField": "django.forms",
+        },
+        "django.contrib.postgres.forms.jsonb": {
+            "JSONField": "django.forms",
+        },
+    },
+}
+REPLACEMENTS_EXCEPT_MIGRATIONS = {
+    (3, 1): {
+        "django.contrib.postgres.fields": {
+            "JSONField": "django.db.models",
+            "KeyTextTransform": "django.db.models.fields.json",
+            "KeyTransform": "django.db.models.fields.json",
+        },
+        "django.contrib.postgres.fields.jsonb": {
+            "JSONField": "django.db.models",
+            "KeyTextTransform": "django.db.models.fields.json",
+            "KeyTransform": "django.db.models.fields.json",
+        },
+    }
+}
+
+
+@lru_cache(maxsize=None)
+def _get_replacements(
+    version: tuple[int, int], looks_like_migrations_file: bool
+) -> Mapping[str, dict[str, str]]:
+    replacements: defaultdict[str, dict[str, str]] = defaultdict(dict)
+    for target_version, target_replacements in REPLACEMENTS_EXACT.items():
+        if target_version <= version:
+            for old_module, rewrite in target_replacements.items():
+                replacements[old_module].update(rewrite)
+
+    if not looks_like_migrations_file:
+        for (
+            target_version,
+            target_replacements,
+        ) in REPLACEMENTS_EXCEPT_MIGRATIONS.items():
+            if target_version <= version:
+                for old_module, rewrite in target_replacements.items():
+                    replacements[old_module].update(rewrite)
+
+    replacements.default_factory = None
```

Comment:
second tweak: setting `default_factory` to `None` makes this act like a regular dict after this point. always a good idea with `defaultdict`.

Reply:
oh nice one! Didn't thought about that.